src/nnet3/nnet-chain-training.cc

```diff
@@ -70,21 +70,40 @@ void NnetChainTrainer::Train(const NnetChainExample &chain_eg) {
 
   NnetComputer computer(nnet_config.compute_config, *computation,
                         *nnet_, delta_nnet_);
+  // no adversarial training on the first minibatch to avoid some stats to be
+  // negative due to the scaling with the negative learning rate
+  if (nnet_config.alpha > 0.0 && num_minibatches_processed_ > 0) {
+    // adversarial training is incompatible with momentum > 0
+    KALDI_ASSERT(nnet_config.momentum == 0.0);
+    // creates a new NnetComputer object
+    NnetComputer computer_adv(nnet_config.compute_config, *computation,
+                              *nnet_, delta_nnet_);
+    // give the inputs to the computer object.
+    computer_adv.AcceptInputs(*nnet_, chain_eg.inputs);
+    computer_adv.Run();
+
+    this->ProcessOutputs(true, chain_eg, &computer_adv);
+    computer_adv.Run();
+    UpdateParamsWithMaxChange(true);
```

Comment:
please document the meaning of the true/false boolean options by declaring a variable
`bool is_adversarial_step = true;`
and calling it with that variable.  And likewise anywhere else you call this.